text/trace/0192-messaging-semantic-conventions-spec.md

```diff
@@ -0,0 +1,443 @@
+more messages on completion.
+
+A "Receive" span covers such calls, and should link
+to the "Create" spans of all messages that are forwarded via the respective
+call. Depending on the use case, "Receive" spans can correlate with "Process"
+spans or other spans modelling processing operations.
+
+> "Receive" spans SHOULD be created for operations of passing messages to the
+> application, when the those operations are initiated by the application code.
+
+##### General considerations for both push-based and pull-based scenarios
+
+The operations modelled by "Deliver" or "Receive" spans do not strictly refer
+to receiving the message from intermediaries, but instead refer to the
+application receiving messages for processing. If messages are fetched from the
+intermediary and forwarded to the application in one go, the whole operation
+might be covered by a "Deliver" or "Receive" span. However, clients might
+pre-fetch messages from intermediaries and cache those messages, and only
+forward messages to the application at a later time. In this case, the
+operation of pre-fetching and caching should not be covered by the "Deliver" or
+"Receive" spans.
+
+Operations covered by "Deliver" or "Receive" can forward zero messages (e. g.
+to notify the application that no message is available for processing), one
+message, or multiple messages (a batch of messages). "Deliver" and "Receive"
+spans should link to the "Create" span of the messages forwarded, thus those
+spans can link to zero, one, or multiple "Create" spans.
+
+> "Deliver" or "Receive" spans MUST NOT be created for messages which are not
+> forwarded to the application, but are pre-fetched or cached by messaging
+> libraries or SDKs.
+>
+> A single "Deliver" or "Receive" span can account for a single message, for
+> multiple messages (in case messages are passed for processing as batches), or
+> for no message at all (if it is signalled that no messages were received).  For
+> each message it accounts for, the "Deliver" or "Receive" span SHOULD link to
+> the "Create" span for the message.
+
+##### Settlement
+
+Messages can be settled in a variety of different ways. In some cases, messages
+are not settled at all (fire-and-forget), in other cases settlement operations
+are triggered manually by the user, in callback scenarios settlement can be
+automatically triggered by messaging SDKs based on return values of callbacks.
+
+A "Settle" spans should be created for every settlement operation, no matter if
+the settlement operation was manually triggered by the user or automatically
+triggered by SDKs. SDKs will in some cases auto-settle messages in
+push-scenarios, when messages are delivered via callbacks. In this case it is
+recommended to create a parent span, so that a "Settle" span will be a sibling
+of the related "Deliver" span.
+
+In cases where it is not feasible to create spans for settlement operations, an
```

Comment:
We discussed that it's a valid implementation decision to use settlement "events" instead of spans even when it is feasible to create spans.
The settlement spans can be very chatty in many cases, and they sometimes contain no duration or acknowledgment which makes them very suitable for a lightweight event on the "deliver" span.

Reply:
I re-formulated this.